src/providers/liveupdates/BasicPubSubClient.hpp

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <pajlada/signals/signal.hpp>
+#include <unordered_set>
+
+#include "common/QLogging.hpp"
+#include "providers/liveupdates/BasicPubSubWebsocket.hpp"
+#include "singletons/Settings.hpp"
+#include "util/DebugCount.hpp"
+#include "util/Helpers.hpp"
+
+namespace chatterino {
+
+/**
+ * This class manages a single connection
+ * that has at most #maxSubscriptions subscriptions.
+ *
+ * You can safely overload the #subscription method
+ * and e.g. add additional heartbeat logic.
+ *
+ * You can use shared_from_this to get a shared_ptr of this client.
+ *
+ * @tparam Subscription see BasicPubSubManager
+ */
+template <typename Subscription>
+class BasicPubSubClient
+    : public std::enable_shared_from_this<BasicPubSubClient<Subscription>>
+{
+public:
+    // The maximum amount of subscriptions this connections can handle
+    const size_t maxSubscriptions;
+
+    BasicPubSubClient(liveupdates::WebsocketClient &websocketClient,
+                      liveupdates::WebsocketHandle handle,
+                      size_t maxSubscriptions = 100)
+        : websocketClient_(websocketClient)
+        , handle_(std::move(handle))
+        , maxSubscriptions(maxSubscriptions)
+    {
+    }
+
+    virtual ~BasicPubSubClient() = default;
+
+    BasicPubSubClient(const BasicPubSubClient &) = delete;
+    BasicPubSubClient(const BasicPubSubClient &&) = delete;
+    BasicPubSubClient &operator=(const BasicPubSubClient &) = delete;
+    BasicPubSubClient &operator=(const BasicPubSubClient &&) = delete;
+
+    virtual void start()
```

Comment:
If start/stop are not called by "users" of the class, it would be nice if they were not public.

For someone implementing a pubsubclient, it would be nicer if they could override `onConnectionEstablished` to perform on-connect actions